modules/bash-commons/src/array.sh

```diff
@@ -81,4 +81,45 @@ function array_prepend {
 
   updated_ary=( "${ary[@]/#/$prefix}" )
   echo ${updated_ary[*]}
-}
\ No newline at end of file
+}
+
+# Applies a functional reduce to an array of arguments. Supports mathmatical expressions only (see man expr)
+# The first argument should be a single quoted string representing a statment to evaluate values with. It 
+# expects the following variables, expanded as:
+#   $1 - the reducer, starting at the first value in the array, the resulting value of the entire expression 
+#        becomes this value on the next call
+#   $2 - the next item in the array
+#
+# Synopsis:
+#
+# array_reduce <expression> <elements>...
+#
+# Examples:
+# 
+# array_reduce '$1 * $2' 1 2 3 4 5
+#   Returns: 120
+#
+# Hint: In the expression string, $1 is initialized to the first element in the array, and $2 starts as the second element.
+function array_reduce {
+    local -r expression="$1"
+    local reducer="$2"
+    shift 2
```

Comment:
`shift` needs to happen twice, otherwise the first element in an array will be both the starting reducer (`$1`) and first "next item" (`$2`), causing off by one.